src/contracts/core/AVSDirectory.sol

```diff
@@ -48,21 +56,201 @@ contract AVSDirectory is
     ) external onlyWhenNotPaused(PAUSED_OPERATOR_REGISTER_DEREGISTER_TO_AVS) {
+        // Assert `operator` is actually an operator.
+        require(
+            delegation.isOperator(operator),
+            "AVSDirectory.registerOperatorToOperatorSets: operator not registered to EigenLayer yet"
+        );
+        // Assert operator's signature has not expired.
+        require(
+            operatorSignature.expiry >= block.timestamp,
+            "AVSDirectory.registerOperatorToOperatorSets: operator signature expired"
+        );
+        // Assert operator's signature `salt` has not already been spent.
+        require(
+            !operatorSaltIsSpent[operator][operatorSignature.salt],
+            "AVSDirectory.registerOperatorToOperatorSets: salt already spent"
+        );
+
+        if (operatorSignature.signature.length != 0) {
+            // Assert signature provided by `operator` is valid.
+            EIP1271SignatureUtils.checkSignature_EIP1271(
+                operator,
+                calculateOperatorSetRegistrationDigestHash({
+                    avs: msg.sender,
+                    operatorSetIds: operatorSetIds,
+                    salt: operatorSignature.salt,
+                    expiry: operatorSignature.expiry
+                }),
+                operatorSignature.signature
+            );
+        }
+
+        // Mutate `operatorSaltIsSpent` to `true` to prevent future respending.
+        operatorSaltIsSpent[operator][operatorSignature.salt] = true;
+
+        // Register `operator` if not already registered.
+        if (avsOperatorStatus[msg.sender][operator] != OperatorAVSRegistrationStatus.REGISTERED) {
+            avsOperatorStatus[msg.sender][operator] = OperatorAVSRegistrationStatus.REGISTERED;
+            emit OperatorAVSRegistrationStatusUpdated(operator, msg.sender, OperatorAVSRegistrationStatus.REGISTERED);
+        }
+
+        // Loop over `operatorSetIds` array and register `operator` for each item.
+        for (uint256 i = 0; i < operatorSetIds.length; ++i) {
+            // Assert avs is on standby mode for the given `operator` and `operatorSetIds[i]`.
+            if (operatorSignature.signature.length == 0) {
+                require(
+                    onStandby[msg.sender][operator][operatorSetIds[i]],
+                    "AVSDirectory.registerOperatorToOperatorSets: avs not on standby"
+                );
+            }
+
+            // Assert `operator` has not already been registered to `operatorSetIds[i]`.
+            require(
+                !isOperatorInOperatorSet[msg.sender][operator][operatorSetIds[i]],
+                "AVSDirectory.registerOperatorToOperatorSets: operator already registered to operator set"
+            );
+
+            // Mutate calling AVS to operator set AVS status, preventing further legacy registrations.
+            if (!isOperatorSetAVS[msg.sender]) isOperatorSetAVS[msg.sender] = true;
```

Comment:
Nit: This doesn't need to be set in every iteration? Just at the end of the function?